table/server/underdb/hive/src/main/java/alluxio/table/under/hive/util/HiveClientPool.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.table.under.hive.util;
+
+import alluxio.Constants;
+import alluxio.resource.CloseableResource;
+import alluxio.resource.DynamicResourcePool;
+import alluxio.util.ThreadFactoryUtils;
+
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.metastore.HiveMetaHookLoader;
+import org.apache.hadoop.hive.metastore.HiveMetaStoreClient;
+import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.apache.hadoop.hive.metastore.RetryingMetaStoreClient;
+import org.apache.hadoop.hive.metastore.api.NoSuchObjectException;
+import org.apache.thrift.TException;
+
+import java.io.IOException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * A pool for hive clients, since hive clients are not thread safe.
+ */
+@ThreadSafe
+public final class HiveClientPool extends DynamicResourcePool<IMetaStoreClient> {
+  private static final ScheduledExecutorService GC_EXECUTOR =
+      new ScheduledThreadPoolExecutor(1, ThreadFactoryUtils.build("HiveClientPool-GC-%d", true));
+  private static final HiveMetaHookLoader NOOP_HOOK = table -> null;
+
+  private final long mGcThresholdMs;
+  private final String mConnectionUri;
+  private final String mHiveDbName;
+  /** This tracks if the db exists in HMS. */
+  private volatile boolean mDbExists = false;
+
+  /**
+   * Creates a new hive client client pool.
+   *
+   * @param connectionUri the connect uri for the hive metastore
+   * @param hiveDbName the db name in hive
+   */
+  public HiveClientPool(String connectionUri, String hiveDbName) {
+    super(Options.defaultOptions()
+        .setMinCapacity(16)
```

Comment:
do you think 16 may be too high a number? we are assuming there will be at least 16 concurrent udb connections?

Reply:
In a way, yes. If a database has more than 16 tables, it would easily need more concurrent hive clients. For example, tpcds has over 20 tables.